File: src/main/java/no/nav/syfo/consumer/rest/TokenConsumer.java
package no.nav.syfo.consumer.rest;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import static java.util.Objects.requireNonNull;
import static org.springframework.http.HttpMethod.GET;
import static org.springframework.http.HttpStatus.OK;

@Slf4j
@Component
public class TokenConsumer {

    private RestTemplate basicAuthRestTemplate;
    private String url;

    public TokenConsumer(RestTemplate basicAuthRestTemplate,
                         @Value("${security-token-service-token.url}") String url) {
        this.basicAuthRestTemplate = basicAuthRestTemplate;
        this.url = url;
    }

    public String getToken() {
        final HttpHeaders headers = new HttpHeaders();
        headers.setContentType(MediaType.APPLICATION_FORM_URLENCODED);

        final String uriString = UriComponentsBuilder.fromHttpUrl(url)
                .queryParam("grant_type", "client_credentials")
                .queryParam("scope", "openid")
                .toUriString();

        final ResponseEntity<Token> result = basicAuthRestTemplate.exchange(uriString, GET, new HttpEntity<>(headers), Token.class);

        if (result.getStatusCode() != OK) {
            throw new RuntimeException("Henting av token feiler med HTTP-" + result.getStatusCode());
        }

        return requireNonNull(result.getBody()).getAccess_token();
    }
}

File: src/main/kotlin/no/nav/syfo/consumer/rest/TokenConsumer.kt
package no.nav.syfo.consumer.rest


import io.ktor.client.*
import io.ktor.client.features.*
import io.ktor.client.request.*
import io.ktor.http.*
import io.ktor.util.*
import kotlinx.coroutines.runBlocking
import no.nav.syfo.behandling.TokenException

class TokenConsumer(private val httpClient: HttpClient, private val url: String) {
    val token: String
        get() {
            var result = ""
            val params = ParametersBuilder()
            params.append("grant_type", "client_credentials")
            params.append("scope", "openid")
            val genUrl = URLBuilder(URLProtocol.HTTPS, url, parameters = params).toString()
            runBlocking {
                try {
                    result = httpClient.get<Token>(genUrl).access_token
                } catch (cause: ClientRequestException) {
                    throw TokenException(cause.response?.status?.value, cause)
                }
            }
            return result
        }
}
